ansilove.1
.\"
.\" Copyright (c) 2011-2020, Stefan Vogt, Brian Cassidy, and Frederic Cambus
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions are met:
.\"
.\"   * Redistributions of source code must retain the above copyright
.\"     notice, this list of conditions and the following disclaimer.
.\"
.\"   * Redistributions in binary form must reproduce the above copyright
.\"     notice, this list of conditions and the following disclaimer in the
.\"     documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
.\" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS
.\" BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
.\" CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
.\" SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
.\" INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
.\" CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
.\" ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
.\" POSSIBILITY OF SUCH DAMAGE.
.\"
.Dd $Mdocdate: August 28 2020 $
.Dt ANSILOVE 1
.Os
.Sh NAME
.Nm ansilove
.Nd ANSI / ASCII art to PNG converter
.Sh SYNOPSIS
.Nm
.Op Fl dhiqrsSv
.Op Fl b Ar bits
.Op Fl c Ar columns
.Op Fl f Ar font
.Op Fl m Ar mode
.Op Fl o Ar file
.Op Fl R Ar factor
.Op Fl t Ar type
.Ar file
.Sh DESCRIPTION
.Nm
is an ANSI / ASCII art to PNG converter, allowing to convert ANSI and
artscene-related file formats into PNG images, supporting ANSI (.ANS),
PCBoard (.PCB), Binary (.BIN), Artworx (.ADF), iCE Draw (.IDF),
Tundra (.TND) and XBin (.XB) formats.
.Pp
It creates size optimized 4-bit PNG files and supports SAUCE (Standard
Architecture for Universal Comment Extensions), 80x25 and 80x50 PC fonts
(including all the 14 MS-DOS charsets), Amiga fonts, and iCE colors.
.Pp
The options are as follows:
.Bl -tag -width 10n
.It Fl b Ar bits
Set to 9 to render 9th column of block characters (default: 8).
.It Fl c Ar columns
Adjust number of columns for ANSI, BIN, and TND files.
.It Fl d
Enable DOS aspect ratio.
.It Fl f Ar font
Select font for supported formats.
.Pp
Valid options for PC fonts are:
.Bl -tag -width Er
.It Ic 80x25
Default (Code page 437)
.It Ic 80x50
80x50 mode (Code page 437)
.It Ic cp737
Greek (Code page 737)
.It Ic cp775
Baltic (Code page 775)
.It Ic cp850
Latin1 (Code page 850)
.It Ic cp852
Latin2 (Code page 852)
.It Ic cp855
Cyrillic (Code page 855)
.It Ic cp857
Turkish (Code page 857)
.It Ic cp860
Portuguese (Code page 860)
.It Ic cp861
Icelandic (Code page 861)
.It Ic cp862
Hebrew (Code page 862)
.It Ic cp863
French Canadian (Code page 863)
.It Ic cp865
Nordic (Code page 865)
.It Ic cp866
Russian (Code page 866)
.It Ic cp869
Greek (Code page 869)
.It Ic terminus
Terminus (Modern font, code page 437)
.El
.Pp
Valid options for Amiga fonts are:
.Bl -tag -width Er
.It Ic amiga
Alias to Topaz, see below
.It Ic microknight
Original MicroKnight version
.It Ic microknight+
Modified MicroKnight version
.It Ic mosoul
Original mO'sOul font
.It Ic pot-noodle
Original P0T-NOoDLE font
.It Ic topaz
Original Topaz Kickstart 2.x version
.It Ic topaz+
Modified Topaz Kickstart 2.x+ version
.It Ic topaz500
Original Topaz Kickstart 1.x version
.It Ic topaz500+
Modified Topaz Kickstart 1.x version
.El
.It Fl h
Show help.
.It Fl i
Enable iCE colors.
.It Fl m Ar mode
Set rendering mode for ANS files.
.Pp
Valid options are:
.Bl -tag -width Ds
.It Ic ced
Black on gray, with 78 columns.
.It Ic transparent
Render with transparent background.
.It Ic workbench
Use Amiga Workbench palette.
.El
.It Fl o Ar file
Specify output filename/path.
.It Fl q
Suppress output messages (quiet).
.It Fl r
Create Retina @2x output file.
.It Fl R Ar factor
Create Retina output file with custom scale factor.
.It Fl t Ar type
Specify input file type.
.Pp
Valid types are:
.Bl -tag -width Ds
.It Ic ans
ANSi (ANSI escape sequences: ANSI X3.64 standard)
.It Ic adf
Artworx format, supporting custom character sets and palettes
.It Ic bin
Binary format (raw memory copy of text mode video memory)
.It Ic idf
iCE Draw format, supporting custom character sets and palettes
.It Ic pcb
PCBoard Bulletin Board System (BBS) own file format
.It Ic tnd
TundraDraw format, supporting 24-bit color mode
.It Ic xb
XBin format, supporting custom character sets and palettes
.El
.It Fl s
Show SAUCE record without generating output.
.It Fl S
If available, use SAUCE info for render options (ex: width).
.It Fl v
Show version information.
.El
.Sh EXAMPLES
To render a file with iCE colors enabled:
.Pp
.Dl $ ansilove -i file.ans
.Pp
To create a Retina @2x output file:
.Pp
.Dl $ ansilove -r file.ans
.Pp
To creates a Retina output file with a custom scale factor:
.Pp
.Dl $ ansilove -R 3 file.ans
.Pp
To only display SAUCE record without generating output:
.Pp
.Dl $ ansilove -s file.ans
To render a file with transparent background:
.Pp
.Dl $ ansilove -m transparent file.ans
.Pp
To render a file using a custom font:
.Pp
.Dl $ ansilove -f amiga file.txt
.Pp
To render a file using a custom font, rendering the 9th column, enabling
iCE colors, and using DOS aspect ratio:
.Pp
.Dl $ ansilove -f 80x50 -b 9 -d -i file.ans
.Sh AUTHORS
.An -nosplit
.Nm
was written by
.An Stefan Vogt ,
.An Brian Cassidy ,
and
.An Frederic Cambus .


ChangeLog
AnsiLove/C 4.1.2 (2020-07-04)

- Define _GNU_SOURCE in CMakeLists.txt to avoid cluttering source files
- No need to set HAVE_SECCOMP to 1, defining it is enough
- Remove some unneeded seccomp related includes
- Validate architectures for seccomp
- Add seccomp support on i386, tested on glibc and musl
- Remove extraneous underscore in #if defined check
- Use __NR_ instead of SYS_ prefix in #if defined checks



AnsiLove/C 4.1.1 (2020-05-25)

- Various documentation tweaks and improvements
- Define timespecsub macro in compat.h if it doesn't exists
- Use CLOCK_REALTIME if CLOCK_MONOTONIC is not available
- Display processing time on exit.
- Fix includes and add forward declarations in various source files



AnsiLove/C 4.1.0 (2020-02-17)

- Add a new '-t' flag (type), allowing to specify input file type and
  override file type detection
- Various documentation tweaks and improvements



AnsiLove/C 4.0.7 (2020-01-18)

- Use OpenBSD style(9) for function prototypes and declarations
- Replace EXIT_SUCCESS by zero in all SAUCE functions
- Do not use typedef for the sauce struct
- Set icecolors to true if defined in SAUCE record ANSiFlags
- Add test cases for most supported formats, missing an IDF file for now
- Sync the description of Retina output capabilities with reality



AnsiLove/C 4.0.6 (2019-12-10)

- Remove extraneous sauceReadFileName() call, read from existing SAUCE record
- Use strdup() when assigning fileName and fext fallback values
- Call memset() to set all record struct fields to zero
- Free SAUCE record and comments once and for all when exiting the program



AnsiLove/C 4.0.5 (2019-11-19)

- Add ENABLE_SECCOMP build option, to allow building seccomp support
  conditionally
- Disable seccomp by default, it needs more testing on non !amd64 platforms
- Use ${CMAKE_INSTALL_BINDIR} instead of hardcoding 'bin'



AnsiLove/C 4.0.4 (2019-10-28)

- Add initial seccomp support to Ansilove
- Move pledge and seccomp initialization before ansilove_init()
- Move conditional includes for 'pledge' and 'strtonum' in compat.h



AnsiLove/C 4.0.3 (2019-08-16)

- Use errx() for displaying error messages
- Print information messages on stdout
- Introduce a version() function to display version information
- Add a -q flag (quiet) to allow silencing output messages



AnsiLove/C 4.0.2 (2019-08-13)

- Remove most examples from the repository, they continue to live on the
  website
- Move the remaining examples in per author directories, and add LICENSE
  files for these artworks, courtesy of their authors



AnsiLove/C 4.0.1 (2019-06-25)

- Fix a few typos in README and ansilove.1
- Simplify display of default font when no font option is specified
- Drop the leading dot when saving input file extension
- Fix the check for readComments() return value



AnsiLove/C 4.0.0 (2019-01-16)

- Switch to using libansilove
- Add a new switch (-d) to toggle DOS aspect ratio
- Add initial support for PabloDraw 24-bit ANSI sequences
- Ansilove now generates only one output file when -r or -R options
  are used
- Display scale factor when -r or -R options are used
- Allow to specify the number of columns for ANSI and Tundra files
- Restrict maximum allowed number of columns to 4096
- Check asprintf() return value and error if allocation fails
- Bugfixes and improvements in the SAUCE parser
- Refactor font selection processing
- Refactor rendering mode processing
- Use OpenBSD style(9) for function prototypes and declaration
- Add missing headers and remove unnecessary ones
- Remove versionInfo(), simply print the header with version information
  when invoked with -v
- Remove showHelp(), show synopsis instead when invoked with -h
- Add a new example ANSi (Thanks Nail)
- Remove -e switch, examples are now listed in the man page instead
- Make synopsis() only display SYNOPSIS, as it should be.
- Add an EXAMPLES section in ansilove.1
- PC fonts for code pages other than cp437 are now named using the cpNNN
  scheme. Albeit not documented anymore, the previously used names will
  keep working



AnsiLove/C 3.0.9 (2018-06-19)

- Add a new example ANSi (Thanks Sylvao)
- Do not use -Werror by default
- Increment maximum ansi sequences length
- Add code to skip PabloDraw 24-bit ANSI sequences



AnsiLove/C 3.0.8 (2017-12-20)

- Adding new example ANSis (Thanks Burps)
- Use asprintf() to create output filenames
- Reflect OpenBSD's pledge() changes
- Add a -R flag allowing to specify a custom retina scale factor



AnsiLove/C 3.0.7 (2017-07-02)

- Print messages to to stderr instead of stdout
- Ensure binary files have an even size
- Better description of supported file formats
- Skip erase in line (EL) sequences (issue reported by Bart Dumon, thanks!)
- Define palettes in config.h
- Use defined palettes instead of hardcoding each color value in the loaders



AnsiLove/C 3.0.6 (2017-04-22)

- Create an output function to unify the way output files are created
- Prevent .png suffix to be appended when using -o flag (*NIX tool
  convention compatibility)



AnsiLove/C 3.0.5 (2017-04-16)

- Free memory allocated in the PCBoard loader
- Free font data where appropriate
- Add a trailing NUL character to inputFileBuffer
- Remove the custom substr function, call strndup directly
- Free seqGrab where necessary
- Use the ternary operator when shitfing cursor position
- Use perror when printing errors
- Remove SUBSTITUTE_BREAK and WRAP_COLUMN_80 options
- Use fstat to get the input file size
- Create a retina function instead of duplicating code in all loaders
- Fix OpenBSD's style(9) offenders in compat functions



AnsiLove/C 3.0.4 (2016-07-28)

- Relicensed under the BSD 2-Clause license
- Input file is now loaded in main, prior to calling format loaders
- Use stat to get input file size instead of custom file size function
- Remove now unused file size function
- Refactoring bits and columns argument value handling



AnsiLove/C 3.0.3 (2016-07-14)

- Adding some missing includes

src/ansilove.c
/*
 * ansilove.c
 * Ansilove 4.1.2
 * https://www.ansilove.org
 *
 * Copyright (c) 2011-2020 Stefan Vogt, Brian Cassidy, and Frederic Cambus
 * All rights reserved.
 *
 * Ansilove is licensed under the BSD 2-Clause License.
 * See LICENSE file for details.
 */

#include <sys/time.h>
#include <ansilove.h>
#include <err.h>
#include <getopt.h>
#include <stdbool.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

#ifdef HAVE_SECCOMP
#include <sys/prctl.h>
#include <linux/seccomp.h>
#include "seccomp.h"
#endif

#include "compat.h"
#include "config.h"
#include "fonts.h"
#include "sauce.h"
#include "strtolower.h"
#include "types.h"

/* prototypes */
static void synopsis(void);
static void version(void);

static void
synopsis(void)
{
	fprintf(stdout, "SYNOPSIS\n"
	    "     ansilove [-dhiqrsSv] [-b bits] [-c columns] [-f font]"
	    " [-m mode] [-o file]\n"
	    "              [-R factor] [-t type] file\n");
}

static void
version(void)
{
	fprintf(stdout, "AnsiLove/C %s - ANSI / ASCII art to PNG converter\n"
	    "Copyright (c) 2011-2020 Stefan Vogt, Brian Cassidy, and "
	    "Frederic Cambus.\n", VERSION);
}

int
main(int argc, char *argv[])
{
	FILE *messages = NULL;

	/* SAUCE record related bool types */
	bool justDisplaySAUCE = false;
	bool fileHasSAUCE = false;
	bool useSAUCEInfo = false;

	int getoptFlag;

	char *input = NULL, *output = NULL;
	char *fileName = NULL;
	char *font = NULL;
	char *type = NULL;
	int filetype = 0;

	struct timespec begin, end, elapsed;

	static struct ansilove_ctx ctx;
	static struct ansilove_options options;

	const char *errstr;

	if (pledge("stdio cpath rpath wpath", NULL) == -1)
		err(EXIT_FAILURE, "pledge");

#ifdef HAVE_SECCOMP
	if (prctl(PR_SET_NO_NEW_PRIVS, 1, 0, 0, 0)) {
		perror("Can't initialize seccomp");
		return EXIT_FAILURE;
	}

	if (prctl(PR_SET_SECCOMP, SECCOMP_MODE_FILTER, &ansilove)) {
		perror("Can't load seccomp filter");
		return EXIT_FAILURE;
	}
#endif

	if (ansilove_init(&ctx, &options) == -1)
		errx(EXIT_FAILURE, "%s", ansilove_error(&ctx));

	while ((getoptFlag = getopt(argc, argv, "b:c:df:him:o:qrR:sSt:v")) != -1) {
		switch (getoptFlag) {
		case 'b':
			options.bits = strtonum(optarg, 8, 9, &errstr);

			if (errstr)
				errx(EXIT_FAILURE, "Invalid value for bits (must be 8 or 9).");

			break;
		case 'c':
			options.columns = strtonum(optarg, 1, 4096, &errstr);

			if (errstr)
				errx(EXIT_FAILURE, "\nInvalid value for columns (must range from 1 to 4096).");

			break;
		case 'd':
			options.dos = true;
			break;
		case 'f':
			font = optarg;
			for (size_t loop = 0; loop < FONTS; loop++) {
				if (!strcmp(fonts[loop], font)) {
					options.font = fontsId[loop];
					break;
				}
			}
			break;
		case 'h':
			synopsis();
			return EXIT_SUCCESS;
		case 'i':
			options.icecolors = true;
			break;
		case 'm':
			if (!strcmp(optarg, "ced")) {
				options.mode = ANSILOVE_MODE_CED;
			} else if (!strcmp(optarg, "transparent")) {
				options.mode = ANSILOVE_MODE_TRANSPARENT;
			} else if (!strcmp(optarg, "workbench")) {
				options.mode = ANSILOVE_MODE_WORKBENCH;
			}
			break;
		case 'o':
			output = optarg;
			break;
		case 'q':
			messages = fopen("/dev/null", "w");
			break;
		case 'r':
			options.scale_factor = 2;
			break;
		case 'R':
			options.scale_factor = strtonum(optarg, 2, 8, &errstr);

			if (errstr)
				errx(EXIT_FAILURE, "Invalid value for retina scale factor (must range from 2 to 8).");

			break;
		case 's':
			justDisplaySAUCE = true;
			break;
		case 'S':
			useSAUCEInfo = true;
			break;
		case 't':
			type = strtolower(optarg);
			break;
		case 'v':
			version();
			return EXIT_SUCCESS;
		}
	}

	if (optind < argc) {
		input = argv[optind];
	} else {
		synopsis();
		return EXIT_SUCCESS;
	}

	argc -= optind;
	argv += optind;

	/* if -q flag was not set, default to stdout */
	if (!messages)
		messages = stdout;

        /* Starting timer */
        clock_gettime(CLOCK_MONOTONIC, &begin);

	/* let's check the file for a valid SAUCE record */
	struct sauce *record = sauceReadFileName(input);

	/* record == NULL also means there is no file, we can stop here */
	if (record == NULL) {
		errx(EXIT_FAILURE, "File %s not found.", input);
	} else {
		/* if we find a SAUCE record, update bool flag */
		if (!strcmp(record->ID, SAUCE_ID)) {
			fileHasSAUCE = true;
		}
	}

	if (!justDisplaySAUCE) {
		/* gather rendering hints from SAUCE */
		if (useSAUCEInfo && fileHasSAUCE) {
			bool usedSAUCE = false;
			if (record->dataType == 1) {
				if (record->fileType == 0 || record->fileType == 1 || record->fileType == 2) {
					options.columns = record->tinfo1;
					if (record->flags & 1)
						options.icecolors = true;
					if ((record->flags & 6) == 4)
						options.bits = 9;
					if ((record->flags & 24) == 8)
						options.dos = true;
					usedSAUCE = true;
				}
				if (record->fileType == 8) {
					options.columns = record->tinfo1;
					usedSAUCE = true;
				}
			}
			if (record->dataType == 5) {
				options.columns = record->tinfo1;
				usedSAUCE = true;
			}
			/* XBIN (dataType == 6) could also use tinfo1 for width, but we trust the XBIN header more */
			/* font info */
			if ((record->dataType == 1 && (record->fileType == 0 || record->fileType == 1 || record->fileType == 2)) || record->dataType == 5 ) {
				if (strcmp(record->tinfos, "IBM VGA") == 0) {
					font = "80x25";
				}
				if (strcmp(record->tinfos, "IBM VGA50") == 0) {
					font = "80x50";
				}
				if (strcmp(record->tinfos, "IBM VGA 437") == 0) {
					font = "80x25";
				}
				if (strcmp(record->tinfos, "IBM VGA50 437") == 0) {
					font = "80x50";
				}
				if (strcmp(record->tinfos, "IBM VGA 775") == 0) {
					font = "baltic";
				}
				if (strcmp(record->tinfos, "IBM VGA50 855") == 0) {
					font = "cyrillic";
				}
				if (strcmp(record->tinfos, "IBM VGA 863") == 0) {
					font = "french-canadian";
				}
				if (strcmp(record->tinfos, "IBM VGA 737") == 0) {
					font = "greek";
				}
				if (strcmp(record->tinfos, "IBM VGA 869") == 0) {
					font = "greek-869";
				}
				if (strcmp(record->tinfos, "IBM VGA 862") == 0) {
					font = "hebrew";
				}
				if (strcmp(record->tinfos, "IBM VGA 861") == 0) {
					font = "icelandic";
				}
				if (strcmp(record->tinfos, "IBM VGA 850") == 0) {
					font = "latin1";
				}
				if (strcmp(record->tinfos, "IBM VGA 852") == 0) {
					font = "latin2";
				}
				if (strcmp(record->tinfos, "IBM VGA 865") == 0) {
					font = "nordic";
				}
				if (strcmp(record->tinfos, "IBM VGA 860") == 0) {
					font = "portuguese";
				}
				if (strcmp(record->tinfos, "IBM VGA 866") == 0) {
					font = "russian";
				}
				if (strcmp(record->tinfos, "IBM VGA 857") == 0) {
					font = "turkish";
				}
				if (strcmp(record->tinfos, "Amiga MicroKnight") == 0) {
					font = "microknight";
				}
				if (strcmp(record->tinfos, "Amiga MicroKnight+") == 0) {
					font = "microknight+";
				}
				if (strcmp(record->tinfos, "Amiga mOsOul") == 0) {
					font = "mosoul";
				}
				if (strcmp(record->tinfos, "Amiga P0T-NOoDLE") == 0) {
					font = "pot-noodle";
				}
				if (strcmp(record->tinfos, "Amiga Topaz 1") == 0) {
					font = " topaz500";
				}
				if (strcmp(record->tinfos, "Amiga Topaz 1+") == 0) {
					font = "topaz500+";
				}
				if (strcmp(record->tinfos, "Amiga Topaz 2") == 0) {
					font = "topaz";
				}
				if (strcmp(record->tinfos, "Amiga Topaz 2+") == 0) {
					font = "topaz+";
				}
				for (size_t loop = 0; loop < FONTS; loop++) {
					if (!strcmp(fonts[loop], font)) {
						options.font = fontsId[loop];
						break;
					}
				}
			}
			if (usedSAUCE) {
				fprintf(messages, "SAUCE info used for rendering hints\n\n");
			}
		}

		/* create output file name if output is not specified */
		if (!output) {
			/* appending ".png" extension to output file name */
			if (asprintf(&fileName, "%s%s", input, ".png") == -1)
				errx(EXIT_FAILURE, "Memory allocation error.");
		} else {
			fileName = strdup(output);
		}

		/* display name of input and output files */
		fprintf(messages, "Input File: %s\n", input);
		fprintf(messages, "Output File: %s\n", fileName);

		/* get file extension */
		char *fext = strrchr(input, '.');
		fext = fext ? strtolower(strdup(++fext)) : strdup("");

		/* check if current file has a .diz extension */
		if (!strcmp(fext, "diz"))
			options.diz = true;

		if (ansilove_loadfile(&ctx, input) == -1)
			errx(EXIT_FAILURE, "%s", ansilove_error(&ctx));

		/* adjust the file size if file contains a SAUCE record */
		if (fileHasSAUCE)
			ctx.length -= 129 - (record->comments > 0 ? 5 + 64 * record->comments : 0);

		int (*loader)(struct ansilove_ctx *, struct ansilove_options *) = NULL;

		/* if type was specified, attempt to find a loader */
		if (type) {
			for (size_t loop = 0; loop < TYPES; loop++) {
				if (!strcmp(types[loop], type)) {
					loader = loaders[loop];
					filetype = filetypes[loop];
					break;
				}
			}

			if (!loader)
				errx(EXIT_FAILURE, "Unknown file type.");
		}

		/* use file extension to find a suitable loader */
		if (!loader) {
			for (size_t loop = 0; loop < TYPES; loop++) {
				if (!strcmp(types[loop], fext)) {
					loader = loaders[loop];
					filetype = filetypes[loop];
					break;
				}
			}
		}

		/* default to ANSI if file extension is unknown */
		if (!loader) {
			loader = ansilove_ansi;
			filetype = ANSILOVE_FILETYPE_ANS;
		}

		if (loader(&ctx, &options) == -1)
			errx(EXIT_FAILURE, "%s", ansilove_error(&ctx));

		/* create the output file */
		if (ansilove_savefile(&ctx, fileName) == -1)
			errx(EXIT_FAILURE, "%s", ansilove_error(&ctx));

		/* gather information and report to the command line */
		switch(filetype) {
		case ANSILOVE_FILETYPE_ANS:
		case ANSILOVE_FILETYPE_BIN:
			if (options.icecolors)
				fprintf(messages, "iCE Colors: enabled\n");

			/* FALLTHROUGH */
		case ANSILOVE_FILETYPE_PCB:
		case ANSILOVE_FILETYPE_TND:
			fprintf(messages, "Font: %s\n", font ? font : "80x25");
			fprintf(messages, "Bits: %d\n", options.bits);
			fprintf(messages, "Columns: %d\n", options.columns);
		}

		if (options.scale_factor)
			fprintf(messages, "Scale factor: %d\n", options.scale_factor);

		free(fileName);
		free(fext);
	}

	/* either display SAUCE or tell us if there is no record */
	if (!fileHasSAUCE) {
		fprintf(messages, "\nFile %s does not have a SAUCE record.\n", input);
	} else {
		fprintf(messages, "\nId: %s v%s\n", record->ID, record->version);
		fprintf(messages, "Title: %s\n", record->title);
		fprintf(messages, "Author: %s\n", record->author);
		fprintf(messages, "Group: %s\n", record->group);
		fprintf(messages, "Date: %s\n", record->date);
		fprintf(messages, "Datatype: %d\n", record->dataType);
		fprintf(messages, "Filetype: %d\n", record->fileType);

		if (record->flags)
			fprintf(messages, "Flags: 0b%d%d%d%d%d%d%d%d\n", record->flags >> 7, record->flags >> 6 & 1, record->flags >> 5 & 1, record->flags >> 4 & 1, record->flags >> 3 & 1, record->flags >> 2 & 1, record->flags >> 1 & 1, record->flags & 1);

		if (record->tinfo1)
			fprintf(messages, "Tinfo1: %d\n", record->tinfo1);

		if (record->tinfo2)
			fprintf(messages, "Tinfo2: %d\n", record->tinfo2);

		if (record->tinfo3)
			fprintf(messages, "Tinfo3: %d\n", record->tinfo3);

		if (record->tinfo4)
			fprintf(messages, "Tinfo4: %d\n", record->tinfo4);

		fprintf(messages, "Tinfos: %s\n", record->tinfos);
		if (record->comments > 0) {
			fprintf(messages, "Comments: ");
			for (int32_t i = 0; i < record->comments; i++)
				fprintf(messages, "%s\n", record->comment_lines[i]);
		}
	}

	/* Stopping timer */
	clock_gettime(CLOCK_MONOTONIC, &end);

	timespecsub(&end, &begin, &elapsed);

	fprintf(messages, "\nProcessed in %f seconds.\n",
	    elapsed.tv_sec + elapsed.tv_nsec / 1E9);

	ansilove_clean(&ctx);

	free(record->comment_lines);
	free(record);

	return EXIT_SUCCESS;
}
